ftms_filtering.qmd

```diff
@@ -360,47 +366,230 @@ print(t <- table(max_level_per_tree$max_level))
+    combined <- top_ms3
+  }
+
+  combined
+} 
+```
+
+### Prepare merged MSntree dataset
+```{r}
+# Extract spectraData from ftms_msn_tree_best
+sd_best <- as.data.frame(spectraData(ftms_msn_tree_best))
+
+sd_filtered$precMzRound <- round(sd_filtered$precursorMz)
+sd_best$precMzRound <- round(sd_best$precursorMz)
+
+# Extract MS4 spectra from sd_filtered
+# ms3_acq: acquisition number of the parent MS3 spectrum
+# ms4_acq: acquisition number of the MS4 spectrum
+# peaksCount.MS4: peaksCOunt of MS4 spectrum
+ms4_spectra <- sd_filtered %>%
+  filter(msLevel == 4) %>%
+  select(MSntreeID, precScanNum, acquisitionNum, peaksCount) %>%
+  dplyr::rename(ms3_acq = precScanNum, ms4_acq = acquisitionNum, peaksCount.MS4 = peaksCount)
+
+
+# Initialize empty list to store additional MS3
+all_additional_ms3 <- list()
+
+for (f in unique(sd_best$feature_id)) {
+  
+  # Best MSn tree for this feature
+  best_tree <- sd_best %>% filter(feature_id == f)
+  best_precMz <- unique(best_tree$precMzRound[best_tree$msLevel == 3])
+  
+  # Candidate MS3 spectra with different precursor m/z
+  ms3_candidates <- sd_filtered %>%
+    filter(feature_id == f,
+           msLevel == 3,
+           !(round(precursorMz) %in% best_precMz))
+  
+  # Only continue if there are candidates
+  if (nrow(ms3_candidates) != 0) {
+    print(paste(f,"has", nrow(ms3_candidates), "additional MS3 candidates.", sep=" " ))
+
+    # Join MS3 candidates to their dependent MS4 (if exists)
+    ms3_candidates <- ms3_candidates %>%
+      left_join(ms4_spectra, by = c("MSntreeID" = "MSntreeID", "acquisitionNum" = "ms3_acq"))
+    
+  
+    # Group by precMzRound, select best MS3, and append MS4
+
+    ms3.p.precmz <- split(ms3_candidates, ms3_candidates$precMzRound)
+    additional.ms3 <- lapply(ms3.p.precmz, selectbest.ms3ms4)
+    all_additional_ms3[[f]] <- do.call(rbind, additional.ms3)
+  }
 }
 
-ftms_one_tree <- unique_tree_feature(ftms_msn_tree)
+# Bind all features into a single data frame
+all_additional_ms3 <- do.call(rbind, all_additional_ms3)
+row.names(all_additional_ms3) <- NULL
+
+
+
+# merge the additional MS3 spectra to the selected best trees
+common_cols <- intersect(names(sd_best), names(all_additional_ms3))
+merged <- bind_rows(
+  sd_best[, common_cols],
+  all_additional_ms3[, common_cols]
+)  
+
+# create new unique merged tree ID 
+merged <- merged %>%
+  mutate(MergedMSntreeID = as.integer(factor(feature_id)))
+
+# Extract and update the corresponding spectra from the original MSn tree object
+mergeKey <- paste(merged$dataOrigin, merged$scanIndex, sep = "_")
+
+original_key <- paste(spectraData(ftms_msn_tree)$dataOrigin,
+                      spectraData(ftms_msn_tree)$scanIndex,
+                      sep = "_")
+
+merged_spectra <- ftms_msn_tree[original_key %in% mergeKey]
+merged <- S4Vectors::DataFrame(merged)
+spectraData(merged_spectra) <- merged
 
-print (ta <- msLevel(ftms_one_tree) |> table())
 ```
 
+## MS4 spectra from other `round(`precursor_mz_MS2)` 
```

Comment:
@rebdau , what is `precursor_mz_MS2`? is this the precursor m/z of a MS3 spectrum (i.e. the ion/precursor selected in a MS2 spectrum)?